ir/src/error.rs

```diff
@@ -74,4 +80,150 @@ impl SemanticError {
+            access.trace_segment(),
+            size
+        ))
+    }
+
+    pub(super) fn indexed_trace_column_access_out_of_bounds(
+        access: &IndexedTraceAccess,
+        segment_width: u16,
+    ) -> Self {
+        SemanticError::IndexOutOfRange(format!(
+            "Out-of-range index '{}' in trace segment '{}' of length {}",
+            access.col_idx(),
+            access.trace_segment(),
+            segment_width
+        ))
+    }
+
+    pub(super) fn random_value_access_out_of_bounds(index: usize, size: u16) -> Self {
+        SemanticError::IndexOutOfRange(format!(
+            "Random value index {index} is greater than or equal to the total number of random values ({size})."
+        ))
+    }
+
+    // --- DECLARATION ERRORS ---------------------------------------------------------------------
+
+    fn missing_declaration(missing_section: &str) -> Self {
+        SemanticError::MissingDeclaration(format!("{missing_section} section is missing"))
+    }
+
+    pub(super) fn missing_trace_columns_declaration() -> Self {
+        Self::missing_declaration("trace_declaration")
+    }
+
+    pub(super) fn missing_public_inputs_declaration() -> Self {
+        Self::missing_declaration("public_inputs")
+    }
+
+    pub(super) fn missing_boundary_constraints_declaration() -> Self {
+        Self::missing_declaration("boundary_constraints")
+    }
+
+    pub(super) fn missing_integrity_constraints_declaration() -> Self {
+        Self::missing_declaration("integrity_constraints")
+    }
+
+    pub(super) fn has_random_values_but_missing_aux_trace_columns_declaration() -> Self {
+        SemanticError::MissingDeclaration(
+            "random_values section requires aux_trace_columns section, which is missing"
+                .to_string(),
+        )
+    }
+
+    pub(super) fn illegal_identifer_type(ident_name: &str, ident_type: &IdentifierType) -> Self {
+        SemanticError::InvalidUsage(format!(
+            "Identifier {ident_name} was declared as a {ident_type} which is not a supported type."
+        ))
+    }
+
+    // --- ILLEGAL IDENTIFIER ERRORS --------------------------------------------------------------
+
+    pub(super) fn duplicate_identifer(
+        ident_name: &str,
+        ident_type: IdentifierType,
+        prev_type: IdentifierType,
+    ) -> Self {
+        SemanticError::DuplicateIdentifier(format!(
+            "Cannot declare {ident_name} as a {ident_type}, since it was already defined as a {prev_type}"))
+    }
+
+    pub(super) fn undeclared_identifier(ident_name: &str) -> Self {
+        SemanticError::InvalidIdentifier(format!("Identifier {ident_name} was not declared"))
+    }
+
+    pub(super) fn illegal_trace_column_identifier(
+        ident_name: &str,
+        ident_type: &IdentifierType,
+    ) -> Self {
+        SemanticError::InvalidUsage(format!(
+            "Identifier {ident_name} was declared as a {ident_type} not as a trace column"
+        ))
+    }
```

Comment:
I think this error gets used when we expect a trace column but instead we get some other type, so I'm not sure that it's an illegal identifier error - it's more of a type mismatch error. I think `illegal_trace_column_identifier` is a bit of a misnomer, and maybe the error message should be rephrased as well. Wdyt?

Reply:
I like `TYPE ERRORS` as a section, and the one where the type of an identifier is illegal can go into that section as well. Does that make sense?

Reply:
Yep, that sounds good to me!